doc/compiler/JProfiling.md

```diff
@@ -20,6 +20,124 @@ OpenJDK Assembly Exception [2].
+to avoid any matches. This is achieved by backing up the values, setting
+the first slot to -1 and setting the rest to 0, relying on that fact
+that a value of 0 will always be placed on the first slot due to the
+hash function. Values are then rearranged in their backed up data structure
+and then committed back once the process has finished. Frequencies can
+be rearranged in place, as no increments should occur.
+
+During this process, other threads will increment the *other* counter so
+that the total frequency remains accurate. This process is complicated by 
+placing in the *other* counter in an unused slot. This is desirable, as a
+table with a capacity of 3 or greater will have at least one counter slot
+that is not in use. The frequency slot being used as the *other* counter
+cannot move during rearragnment, as it may race with other moves. To
+get around this, the final destination of the *other* counter is stashed
+and then performed as the final rearrangement.
+
+Once the values and frequencies have been rearranged, the new hash
+configuration can be committed and the addition is finished.
+If the table reached its capacity, the full flag is set and any new
+values seen in jitted code are attributed to the *other* counter.
+An additional optimization is made here, to cut down on memory accesses.
+The table is considered full when the index for the *other* counter
+is positive, so the tested value can be used immediately to increment
+the counter. In the event that its negative, the helper should be
+called to add a new value to the table.
+
+As the table populates based on the first distinct values that are
+seen, its possible to have ignored highly frequent values, for example due to
+an early phase change. To account for this, a thread inspects the hash
+tables and clears any with high *other* counters relative to that of the
+profiled values. If it decides that highly frequency values have been
+missed, it will reset the table and allow it to repopulate.
+
+Each table has a lock bit, which is locked when accessing data from a
+compilation thread, rearranging in the helper or clearing in the
+thread. Increments from the mainline do check this lock, to avoid
+the overhead. Additionally, these increments are not atomic, due to
+the overhead, which may result in races. In testing, this did not affect
+results sufficiently to negatively impact optimizations.
+It should be noted that, unlike the global monitor used by other profiling
+implementations, the lock is not reentrant.
+
+In addition to the `pext` hash function, other are supported. One approach
+uses an array of bytes specifying shifts to apply to the original value.
+This is implemented using existing opcodes and should be supported on all
+platforms. A more efficient implementation exists using the new `bitpermute`
+opcode, configured with an array of bytes specifying bit indices, rather
+than shifts. If supported by the codegen, the `bitpermute` approach will
```

Comment:
perhaps name the query which must return true from the cg for the bitpermute to be used?